Consistent built-in (#3600)
modelica/ModelicaSpecification: chapters/operatorsandexpressions.tex

\end{lstlisting}
\end{example}

\section{Built-in Operators and Functions}\label{built-in-intrinsic-operators-with-function-syntax}

Certain built-in operators of Modelica are called using the same syntax as a function call.
However, they do not behave as a mathematical function, because the result depends not only on the input arguments but also on the status of the simulation.

There are also built-in functions that depend only on the input argument, but also may trigger events in addition to returning a value.
The built-in functions may also be overloaded such that a single Modelica function cannot be compatible with all calls of the function.

> Built-in means that they are defined at the Modelica language level, not in the Modelica library.

Reviewer: ALso, shouldn't it be _a Modelica library_ or _the Modelica Standard Library_?

> Built-in means that they are defined at the Modelica language level, not in the Modelica library.

Author: I would prefer the more general 'a', so that people don't think they might define in some other library.
(Using 'any' seemed to general.)

The following built-in operators/functions are available:
\begin{itemize}
\item
  Mathematical functions and conversion operators, see \cref{numeric-functions-and-conversion-functions} below.
\item
  Derivative and special purpose operators with function syntax, see \cref{derivative-and-special-purpose-operators-with-function-syntax} below.
\item

>   Event-related functions, see \cref{event-related-operators-with-function-syntax} below.

Reviewer: But the event-generating stuff is not following the variability rules of normal function calls, so they should be operators, right?
```suggestion
  Event-related operators, see \cref{event-related-operators-with-function-syntax} below.
```

Author: After we added GenerateEvents-annotations I think they can be seen as following the rules.
(With the caveat that the rules are slightly unclear.)

Reviewer: Yeah, we haven't figured out those variability rules yet…

\item

>   Array functions, see \cref{built-in-array-functions}.

Reviewer: Many of the built-ins for arrays are operators.  For instance, a function couldn't describe the result type of `promote`, and `ndims` and `size` don't follow the variability rules for functions.  If anything, I think that it is the title of the section that should be changed (and that some of the definitions in the chapter should use `functiondefinition` instead of `operatordefinition`).
```suggestion
  Array operators/functions, see \cref{built-in-array-functions}.
```

\item
  Synchronous operators, see \cref{synchronous-language-elements}.
\item
  State machine operators, see \cref{transitions}.
\end{itemize}

> Note that when the specification references a function or operator having the name of a built-in function or operator it references the built-in function or operator, not a user-defined function having the same name, see also \cref{built-in-functions}.

Reviewer: This sentence is just got even harder to digest.  Could we simplify it?
```suggestion
Except where shadowing problems are being discussed, references to built-in functions and operators within this document always assume that the built-in definitions are not shadowed by user-defined definitions, see also \cref{built-in-functions}.
```

E.g., \lstinline!noEvent(integer(v))!.
\end{nonnormative}

> \begin{functiondefinition}[div]

Reviewer: As explained above, I don't think `div` is a function.  (Similarly for the other operators in this section.)

Author: To me the annotation `GenerateEvents` should allow you to write a function such as `div` in Modelica.

Reviewer: This has been addressed.  Resolving.
